### PR TITLE
feat: sandbox layer, initial implementation

### PR DESCRIPTION
**Pull request by h2non**
Work in progress!

### CODE REVIEW

**Comment by h2non** [comment]
Hi @groyoh. If you're open to, feel free to take a look to the progress. It's pretty much a work in progress, design-code driven status, but I guess you should be able to intuit in which direction things are moving.

Just to give a bit of perspective, the is related to design the required abstractions and entities that will power vinxi's configuration capabilities. This abstraction mixes concepts of the middleware layer and multiplexer, but with domain specific interfaces.

### Configuration entities

For now I've defined 3 different entities to manage vinxi servers:

#### Scope

Scope represents an generic HTTP scope level of configuration. It's moslty analog to a multiplexer matcher. Scopes are defined by a set of rules and can host a set of plugins to provide specific behavior. 

An example of a scope could be when a user want to enable certain plugins based on certain conditions, such as, specific HTTP verb, hostname, URL path, header, client IP...

#### Rule

Rule represents a simple logic who inspect an incoming HTTP request and determine if the configured rule is passed. It's used to compose and reuse matching logic across scopes and allows to enable plugins based on the rule resolution state.

#### Plugin

Plugin is pretty much as an standard middleware with an additional decorator API to provide plugin specific logic (e.g: enable, disable, remove...). Plugins are in general a simple wrapper with a concrete interface of vinxi components, such as: forward, replay, intercept, metrics...

### HTTP API

I'm starting defining the HTTP API. For now these are the endpoint I've in mind:

/health -> Heart beat endpoint
/config -> Return supported config entities (useful for API consumers)
/config/plugins -> List of supported plugins (useful for API consumers configuration)
/config/scopes -> List supported scopes (useful for API consumers configuration)
/config/rules -> List supported scopes (useful for API consumers configuration)
/plugins -> List, register or remove generic plugins that will be always executed
/scopes -> List, register, remove server configuration scopes
/scopes/:scope -> List current scope details
/scopes/:scope/rules -> List, register, remove scope specific rules
/scopes/:scope/plugins -> List and register scope specific plugins

Example JSON response when requesting `GET /scopes`:

``` json
[
  {
    "id": "fzsDlXgWJjmFNa12",
    "name": "default",
    "rules": [
      {
        "id": "zpqYvl2NTv8qKRgV",
        "name": "path",
        "description": "Matches HTTP request URL path againts a given path pattern",
        "config": "{\"path\": \"\/foo\"}"
      }
    ],
    "plugins": [
      {
        "id": "RJsMM6XmeFh62bYE",
        "name": "static",
        "description": "serve static files",
        "enabled": true
      }
    ]
  }
]
```

**Comment by groyoh** [comment]
I just got back from my holidays today. I'll have a look at it tomorrow :wink:

**Comment by h2non** [comment]
@groyoh great, I hope you really enjoyed them

File: manager.go [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+	rules    []Rule
+	plugins  *PluginLayer
+
+	ID          string
+	Name        string
+	Description string
+}
+
+func NewScope(rules ...Rule) *Scope {
+	return &Scope{ID: uniuri.New(), Name: "default", plugins: NewPluginLayer(), rules: rules}
```
Comment (groyoh) [inline manager.go:43]: How would you specify a different name?

Comment (groyoh) [inline manager.go:43]: If you create two scopes, you would have two `"default"` scopes, right? Would that make sense?

File: manager.go [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+	ID() string
+	Name() string
+	Description() string
+	// Options() Options
+	JSONConfig() string
+	Match(*http.Request) bool
+}
+
+type Scope struct {
+	disabled bool
```
Comment (groyoh) [inline manager.go:33]: Any reason why you chose `disabled` instead of `enabled`?

File: manager.go [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+	return m
+}
+
+func (m *Manager) NewScope(rules ...Rule) *Scope {
+	scope := NewScope(rules...)
+	m.scopes = append(m.scopes, scope)
+	return scope
+}
+
+func (a *Manager) HandleHTTP(w http.ResponseWriter, r *http.Request, h http.Handler) {
```
Comment (groyoh) [inline manager.go]: `a *Manager` ?

File: manager.go [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+
+func (a *Manager) ServeAndListen(opts ServerOptions) (*http.Server, error) {
+	a.Server = NewServer(opts)
+
+	m := pat.New()
+	a.Server.Handler = m
+
+	// Define route handlers
+	m.Get("/", http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		io.WriteString(w, "vinxi HTTP API manager "+Version)
```
Comment (groyoh) [inline manager.go:142]: Where does `Version` come from?

**Comment by h2non** [comment]
Thanks for the review. I'll come back to you with the answers soon. By the way, the code pretty much an experimental thing, I believe it's better to focus on high level functionalities instead of implementation details, which are at this stage very prone to change.

What do you think about the config entities? What do you think about current API and plugin registering approach? Do you feel like this new API is redundant compared to raw middleware layer? What do you could change of the current design and entities, or its responsabilities?

File: manager.go (hunk 4 above)
Comment (h2non) [inline manager.go:142]: version.go at the repository root

File: manager.go (hunk 1 above)
Comment (h2non) [inline manager.go:43]: This is just a temporal thing. My idea is to specify the name via factory function.

File: manager.go (hunk 2 above)
Comment (h2non) [inline manager.go:33]: Simple reason: bool type is `false` state by default. Also, most common scenario comes first for me, so the most common scenario is to have the plugin enabled, unless you explicitely disable it.

File: manager.go (hunk 3 above)
Comment (h2non) [inline manager.go]: Will fix. Code style is not too much relevant now, but thanks for highlighting this.

File: manager.go (hunk 2 above)
Comment (groyoh) [inline manager.go:33]: Alright, makes sense. Just felt unnatural though.

**Comment by groyoh** [comment]
Yeah, at the moment it all makes sense. It does not feel that redundant and it's really well thought. The plugin registration feel natural to me. I wouldn't change anything on this side.

One thing that I can say is that `scope` was not really clear to me at first. After another thought it makes sense but it may be misleading. I would have expected something like `group` or maybe something else.  But maybe that's just me.

**Comment by h2non** [comment]
Scope sounds a bit ambiguous to me. The term could be changed without too much trouble anytime. The relevant part for me to discuss are its responsabilities and API design.

**Comment by h2non** [comment]
Any opinion about the HTTP API proposal?

**Comment by groyoh** [comment]
Yeah, the responsibilities are pretty clear. On the API design:

> /health -> Heart beat endpoint

No heart beat available without using the Manager?

> /config -> Return supported config entities (useful for API consumers)

What are exactly the supported config entities? All plugins, scope and rules?

> /config/plugins -> List of supported plugins (useful for API consumers configuration)
> /config/scopes -> List supported scopes (useful for API consumers configuration)
> /config/rules -> List supported scopes (useful for API consumers configuration)

What do you mean by "supported"? Where would it be defined that they are supported? Also this feels unnatural to prefix it by "config". I would have expected to find the actual configuration instead of "suppported" configuration.

> /plugins -> List, register or remove generic plugins that will be always executed
> /scopes -> List, register, remove server configuration scopes
> /scopes/:scope -> List current scope details
> /scopes/:scope/rules -> List, register, remove scope specific rules
> /scopes/:scope/plugins -> List and register scope specific plugins

This last part sounds good to me. :+1: I would expect to be able to find a scope by name though. What do you think about this?

We might as well only return the scope with every rules and plugins (or is it what you mean by "details"). I don't see the point of listing a scope without the rules and plugin.

**Comment by h2non** [comment]
@groyoh What do you think about merging this repository into the `vinxi` main repository? Possibly as subpackage.

**Comment by groyoh** [comment]
@h2non sounds good to me

**Comment by h2non** [comment]
Ok, sounds good, will do asap. IMO the `vinxi` core package should provide the minimal API to setup a full featured proxy with HTTP API.

**Comment by h2non** [comment]
I'm going to merge this and move the code to `vinxi` main repo.

**Comment by groyoh** [comment]
Yeah, actually at some point I wondered if we shouldn't have every component as sub directories using a monorepo strategy.

**Comment by h2non** [comment]
Well, I throught about this as well, but at the at Go package level would be the same, only with larger import paths.

My proposal is an intermediate scenario where only core packages are in the same repository. I'm especifically talking about the following packages:
- layer
- router
- context

**Pull request merged**